src/sparseml/exporters/transforms/utils/matching.py

```diff
@@ -0,0 +1,329 @@
+    matched against the corresponding entry in node.input/node.output. I.e.
+        - parent_ops[0] will be compared against node.input[0]
+        - parent_ops[2] will be compared against node.input[2]
+        - children_ops[3] will be compared against node.output[3]
+    Because of this, if the length of node.input/node.output is shorter
+    than parent_ops/children_ops, it is not a possible match and will
+    be discarded.
+
+    Here's a simple example of matching against an identity node
+    with a single parent (because there is a single list in parent_ops),
+    and the parent must be an initializer (the INITIALIZER_MATCH value):
+    ```python
+    matches = get_structural_matches(
+        graph,
+        op_type="Identity",
+        parent_ops=[[INITIALIZER_MATCH]]
+    )
+    for match in matches:
+        id_node = match.node
+        (init, ) = match.parents[0]
+        assert isinstance(init, onnx.TensorProto)
+    ```
+
+    Here is a really complicated example with optional nodes and multiple parents. Here
+    we match against MatMul nodes that have at least 2 inputs from Quant/Dequant
+    sequences. After the MatMul node there are two optional nodes followed by
+    a Quant/Dequant sequence.
+
+    If the optional_nodes are found, then the values in children_ops will be the nodes.
+    If the optional nodes are NOT found, then those entries in children_ops
+    will be None.
+
+    In both cases, the length of match.children[0] will be the same as
+    the length of children_ops[0].
+
+    ```python
+    matches = get_structural_matches(
+        graph,
+        op_type="MatMul",
+        parent_ops=[
+            ["QuantizeLinear", "DequantizeLinear"],
+            ["QuantizeLinear", "DequantizeLinear"],
+        ],
+        children_ops=[
+            [
+                optional_node("Transpose"),
+                optional_node("Reshape"),
+                "QuantizeLinear",
+                "DequantizeLinear",
+            ]
+        ]
+    )
+    for match in matches:
+        matmul_node = match.node
+        parent_0_quant, parent_0_dequant = match.parents[0]
+        parent_1_quant, parent_1_dequant = match.parents[1]
+        opt_transpose, opt_reshape, child_quant, child_dequant = match.children[0]
+    ```
+    """
+
+    # NOTE: gather matches completely first, so we don't have to worry about
+    # updates to the graph messing with the iteration
+    matches = []
+    for node in graph._model.graph.node:
+        match = _match_structure(graph, node, op_type, parent_ops, children_ops)
+        if match is not None:
+            matches.append(match)
+    return matches
+
+
+NodeOrInit = Union[NodeProto, TensorProto]
+"""
+Represents either:
+1. a node (NodeProto) in an onnx graph
+2. an initializer (TensorProto)
+"""
+
+
+class MatchResult:
+    """
+    The output of :func:`get_structural_matches`
+    """
+
+    def __init__(self, node: NodeOrInit) -> None:
```

Comment:
not necessary according to pep
```suggestion
    def __init__(self, node: NodeOrInit):
```